File: processing/syntactic/expressions/macro_application_expr.py
from __future__ import annotations

from typing import Optional

from processing.syntactic.expressions.definitions.macro_def import MacroDef
from processing.syntactic.expressions.expr import Expr, child_property
from processing.syntactic.expressions.macro_patterns import CascadePattern
from processing.syntactic.expressions.macro_patterns import TokenPattern


class MacroApplicationExpr(Expr):
    """Actually, it's not an expression,
       but some piece of code defined as expression
       by language macros."""

    @child_property
    def macro_definition(self) -> MacroDef:
        pass

    def __init__(self, parent: Expr):
        """Tries to match next piece of source
           with previously defined macro syntax-es.
           This class instance should be deleted immediately if it's
           [macro_definition] is null."""
        super().__init__(parent)
        self.macro_definition = None
        self.expressions = None

    def parse(self) -> Optional[MacroApplicationExpr]:
        try:
            self.macro_definition = list(filter(
                lambda m: isinstance(m.syntax.patterns[0], TokenPattern)
                          and m.syntax.match(self),
                self.ast.macros
            ))[0]
        except IndexError:
            self.parent = None
            return None

        if self.macro_definition is not None:
            self.expressions = list(self.ast.macro_application_parts)
            self.ast.macro_application_parts.clear()
        return self

    def parse_infix_macro(self, left_expr: Expr) -> Optional[MacroApplicationExpr]:
        """Infix macros (starts with expression).
           Tries to match next piece of source
           with previously defined macro syntax-es.
           This class instance should be deleted immediately if it's
           [macro_definition] is null."""
        try:
            self.macro_definition = list(filter(
                lambda m: len(m.syntax.patterns) > 0
                          and isinstance(m.syntax.patterns[1], TokenPattern)
                          and m.syntax.patterns[1].value == self.stream.peek.value,
                self.ast.macros
            ))[0]
        except IndexError:
            self.parent = None
            return None

        idx = self.stream.token_idx
        self.stream.eat_any()
        self.ast.macro_application_parts.append(self.stream.token)
        rest_cascade = CascadePattern(self.macro_definition.syntax.patterns[2:])
        if rest_cascade.match(self):
            self.expressions = [left_expr] + self.ast.macro_application_parts
            self.ast.macro_application_parts.clear()
        else:
            self.macro_definition = None
            self.stream.move_abs(idx)
        return self

File: Axion.Core/Processing/Syntactic/Expressions/MacroApplicationExpr.cs
using System.Collections.Generic;
using System.Linq;
using Axion.Core.Processing.Syntactic.Expressions.Common;
using Axion.Core.Processing.Syntactic.Expressions.Definitions;
using Axion.Core.Processing.Syntactic.Expressions.Patterns;

namespace Axion.Core.Processing.Syntactic.Expressions {
    /// <summary>
    ///     Actually, it's not an exact expression,
    ///     but some piece of code defined as expression
    ///     by language macros.
    /// </summary>
    public class MacroApplicationExpr : AtomExpr {
        public MacroDef? Macro { get; set; }

        public List<Node> Expressions { get; } = new List<Node>();

        public MacroApplicationExpr(Node parent) : base(parent) { }

        public MacroApplicationExpr Parse() {
            Ast.MacroApplicationParts.Push(this);
            Macro = Ast.Macros.FirstOrDefault(
                m => m.Syntax.Patterns.Count > 0
                  && m.Syntax.Patterns[0] is TokenPattern
                  && m.Syntax.Match(this)
            );
            if (Macro != default) {
                Start = Expressions[0].Start;
                End   = Expressions[^1].End;
            }
            else {
                Expressions.Clear();
            }
            Ast.MacroApplicationParts.Pop();
            return this;
        }

        public MacroApplicationExpr Parse(Expr leftExpr) {
            Ast.MacroApplicationParts.Push(this);
            var startIdx = Stream.TokenIdx;

            Macro = Ast.Macros.FirstOrDefault(
                m => m.Syntax.Patterns.Count > 1
                  && m.Syntax.Patterns[1] is TokenPattern t
                  && t.Match(this)
            );
            if (Macro != null) {
                var restCascade = new CascadePattern(this) {
                    Patterns = new NodeList<Pattern>(
                        this,
                        Macro.Syntax.Patterns.Skip(2)
                    )
                };
                Expressions.Insert(0, leftExpr);
                if (restCascade.Match(this)) {
                    Start = Expressions[0].Start;
                    End   = Expressions[^1].End;
                }
                else {
                    Macro = null;
                    Stream.MoveAbsolute(startIdx);
                    Expressions.Clear();
                }
            }

            Ast.MacroApplicationParts.Pop();
            return this;
        }
    }
}
